### PR TITLE
Chart

### PR DESCRIPTION
**Pull request by benma**

### CODE REVIEW

**Comment by x1ddos** [comment]
Would be cool to rebase. All other backend/rates pending changes are merged.
There's also RateUpdater.HistoryEarliestTimestamp and `HistoryLatestTimestamp` to peek at when the data is ready.

**Comment by benma** [comment]
Rebased, but unfinished. Still need to check if data is available to show the error message (currently the chart can just be 0 where the price info is not available), and some other things.

**Comment by benma** [comment]
@x1ddos I force pushe. Now the chart is shown in daily points until `now - 3 months` and hourly points after.

I am using the `HistoryEarliestTimestamp()` and `HistoryLatestTimestamp()` functions to determine if we should show the chart or show the 'gathering info' error.

To be fixed in a separate PR: the coingecko rate limiter, combined with the fact that we start fetching the history for *all* coins instead of only the active ones, means I am waiting for any chart data for quite a long time if I only have e.g. BTC+LTC active. 

PTAL.

**Comment by x1ddos** [comment]
> the fact that we start fetching the history for all coins instead of only the active ones

wait, it should be the opposite. we fetch historical data only for active coins: https://github.com/digitalbitbox/bitbox-wallet-app/blob/8674f8f/backend/backend.go#L209-L225

unless I misunderstood.

**Comment by benma** [comment]
> unless I misunderstood.

Indeed ;) The active coins in the settings page are not equal to the coins/accounts loaded in the sidebar. If you have a BTC-only device, the settings could have all ERC20 tokens enabled, and the rates are being fetched, but never used.

Maybe not a huge deal anyway, let's test first before improving it.

File: backend/handlers/handlers.go [hunk 1]
```diff
@@ -840,16 +840,119 @@ func (handlers *Handlers) getAccountSummary(_ *http.Request) (interface{}, error
...
+		)
+
+		for i, e := range timeseries {
+			price := handlers.backend.RatesUpdater().PriceAt(
+				string(account.Coin().Code()),
+				fiat,
+				e.Time)
+			timestamp := e.Time.Unix()
+			if chartEntries[i].Time != 0 && chartEntries[i].Time != timestamp {
+				panic("expected same timestamp for all data entries")
```
Comment (x1ddos) [inline backend/handlers/handlers.go:921]: I'd rather confirm this doesn't happen with a test than panic.

**Review by x1ddos — APPROVED** [review]
LGTM for a first version of MVP. A couple TODOs but ok to do it in follow up commits. Maybe @thisconnect wants to check this out.

> Indeed ;) The active coins in the settings page are not equal to the coins/accounts loaded in the sidebar. If you have a BTC-only device, the settings could have all ERC20 tokens enabled, and the rates are being fetched, but never used.

So, that's an issue specific to BTC-only device, right.

File: backend/handlers/handlers.go [hunk 2]
```diff
@@ -840,16 +840,119 @@ func (handlers *Handlers) getAccountSummary(_ *http.Request) (interface{}, error
...
+		if earliestPriceAvailable.IsZero() ||
+			latestPriceAvailable.IsZero() ||
+			(!earliestTxTime.IsZero() && earliestTxTime.Before(earliestPriceAvailable)) ||
+			until.After(latestPriceAvailable) {
+			chartDataMissing = true
+			continue
+		}
+
+		timeseriesDaily, err := txs.Timeseries(
+			time.Now().AddDate(-4, 0, 0).Truncate(24*time.Hour).Add(time.Hour),
```
[the remark this answers is not in the document]
Comment (benma) [reply backend/handlers/handlers.go:876]: I think I'll resolve this later. Should be the time of the earliest tx I think.

**Pull request merged**